### ISSUE TITLE
ClassOrdering false positive - inner/local extension var/val order

### ISSUE THREAD

**Issue by zayo** [detekt/detekt#3532]
Hi,

in our project, we tend to help ourselves with lot of private extension methods that serves the better readability.

For example
```kotlin
class Helper(private val intent: Intent) {
    private val isConnected: Boolean
        get() = !intent.isNoConnectivity

    public fun getPendingIntent(): PendingIntent {
          return intent.toPendingIntent()
    }
    
    private val Intent.isNoConnectivity: Boolean get() =
        getBooleanExtra(ConnectivityManager.EXTRA_NO_CONNECTIVITY, false)
}
```

For us, this (Intent.isNoConnectivity) is just idiomatic way of private methods that would otherwise mess up the code flow.

However, the new rule **ClassOrdering** forces these extensions to be ordered/mixed with private properties of target class. I believe it's a false positive, and every extension var/val should act more like method, taking in account only public/private at most, or be ignored completely.

* Version of detekt used: 1.16.0-RC3

**Comment by chao2zhang** [comment]
Linking to the ktlint discussion (https://github.com/pinterest/ktlint/issues/309) on the difficulty of formatting tools to orchestrate class layout.

This request seems to go beyond the existing [Kotlin style guide](https://kotlinlang.org/docs/coding-conventions.html#class-layout), would you mind submitting a PR to detail your proposed changes to the guidelines? I am a little bit hesitant to modify the existing implementation of `ClassOrdering`.

**Comment by cortinico** [comment]
> However, the new rule ClassOrdering forces these extensions to be ordered/mixed with private properties of target class

I don't see how extension fun/val should be considered an exception. You could also have a `private fun` that is logically coupled to a `public` fun/val and you want to keep it locally near.

The rationale behind this rule is to make sure your public API is ordered and appears on top of the class and not spread between `private` methods.

**Comment by zayo** [comment]
In fact, the extension properties does not exist. You cannot modify the receiver object, nor append any field, as doc says.
In other words, the extension property inside class serves purpose of public/private fun, which means, they should be treated as public/private fun.

But they are expected to be sorted as private fields, which the are not.
I've edited original post, as this is not true extension funs, they are treated as normal funs. So only problematic part is the extension vals/vars

**Comment by BraisGabin** [comment]
So you are talking that the private extension properties should not be treated as properties but as private functions, right?  As @chao2zhang said this is a really difficult topic. But I would approve a PR doing that change, a extension property is just a function (or two) but hidden behind some syntax sugar so it should be treated as a function.

Do you have the time to create a PR changing this?

**Comment by zayo** [comment]
@BraisGabin I tried. My first time, hope I didn't do anything wrong :)
To your comment, I don't mean this is necessarily problem of only private/public val/var.

In fact this issue is raised by the fact that extension function is defined inside class. That is because that property is using the internal state of the class. But that doesn't necessarily mean, it has to be private. It can be as well public.
Example could be optional fields. Something you would normally do like `open val color: Color? = null`. But that is messing the API with `color` property even for subclasses for which this might make no sense.
It's definitely rare, but there are cases when even public makes sense (this mine might be over constructed).

So that's just the explanation, why I did not care about visibility.

**Comment by schalkms** [comment]
I suggest to close this issue, since there's nothing actionable at the moment. If the PR in the kotlin doc repo gets approved, we could reopen this issue.
